ethcore/private-tx/src/log.rs

```diff
@@ -0,0 +1,247 @@
+		}
+		logs.insert(tx_hash, TransactionLog {
+			tx_hash,
+			status: PrivateTxStatus::Created,
+			creation_timestamp: SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs(),
+			validators: validator_logs,
+			deployment_timestamp: None,
+			public_tx_hash: None,
+		});
+	}
+
+	/// Logs the obtaining of the signature for the private transaction
+	pub fn signature_added(&self, tx_hash: H256, validator: Address) {
+		let mut logs = self.logs.write();
+		if let Some(transaction_log) = logs.get_mut(&tx_hash) {
+			transaction_log.status = PrivateTxStatus::Validating;
+			if let Some(ref mut validator_log) = transaction_log.validators.iter_mut().find(|log| log.account == validator) {
+				validator_log.validated = true;
+				validator_log.validation_timestamp = Some(SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs());
+			}
+		}
+	}
+
+	/// Logs the final deployment of the resulting public transaction
+	pub fn tx_deployed(&self, tx_hash: H256, public_tx_hash: H256) {
+		let mut logs = self.logs.write();
+		if let Some(log) = logs.get_mut(&tx_hash) {
+			log.status = PrivateTxStatus::Deployed;
+			log.deployment_timestamp = Some(SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs());
+			log.public_tx_hash = Some(public_tx_hash);
+		}
+	}
+
+	fn read_logs(&self) {
+		let log_file = match self.logs_dir {
+			Some(ref path) => {
+				let mut file_path = path.clone();
+				file_path.push("private_tx.log");
+				match File::open(&file_path) {
+					Ok(file) => file,
+					Err(err) => {
+						trace!(target: "privatetx", "Cannot open logs file: {}", err);
+						return;
+					}
+				}
+			}
+			None => {
+				warn!(target: "privatetx", "Logs path is not defined");
+				return;
+			}
+		};
+		let mut transaction_logs: Vec<TransactionLog> = match serde_json::from_reader(log_file) {
+			Ok(logs) => logs,
+			Err(err) => {
+				error!(target: "privatetx", "Cannot deserialize logs from file: {}", err);
+				return;
+			}
+		};
+		// Drop old logs
+		let current_timestamp = SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs();
+		transaction_logs.retain(|tx_log| current_timestamp - tx_log.creation_timestamp < MAX_STORING_TIME);
+		let mut logs = self.logs.write();
+		for log in transaction_logs {
+			logs.insert(log.tx_hash, log);
+		}
+	}
+
+	fn flush_logs(&self) {
+		if self.logs.read().is_empty() {
+			// Do not create empty file
+			return;
+		}
+		let log_file = match self.logs_dir {
+			Some(ref path) => {
+				let mut file_path = path.clone();
+				file_path.push("private_tx.log");
+				match File::create(&file_path) {
+					Ok(file) => file,
+					Err(err) => {
+						trace!(target: "privatetx", "Cannot open logs file for writing: {}", err);
+						return;
+					}
+				}
+			}
+			None => {
+				return;
+			}
+		};
+		if let Err(err) = serde_json::to_writer(log_file, &self.logs.read().values().cloned().collect::<Vec<TransactionLog>>()) {
```

Comment:
Instead of cloning the values, how about:

```rust
let logs = self.logs.read();
let mut json = serde_json::Serializer::new(log_file);
let mut json_array = json.serialize_seq(Some(logs.len()))?;
for v in logs.values() {
	json_array.serialize_element(v)?;
}
json_array.end()?;
```